File: kdeui/src/part/packetmanager.cpp

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  KDE User Interface                                                    *
 *                                                                        *
 *  Copyright (c) 1999-2011, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public             *
 *  License along with this program; if not, write to the Free            *
 *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,       *
 *  MA 02110-1301, USA.                                                   *
 *                                                                        *
 **************************************************************************/

/* end stub */

// Regina core includes:
#include "packet/packetregistry.h"
#include "surfaces/filterregistry.h"

// UI includes:
#include "packetmanager.h"
#include "reginapart.h"
#include "reginasupport.h"
#include "packettypes/nanglestructureui.h"
#include "packettypes/ncontainerui.h"
#include "packettypes/nnormalsurfaceui.h"
#include "packettypes/npdfui.h"
#include "packettypes/nscriptui.h"
#include "packettypes/nsurfacefiltercomb.h"
#include "packettypes/nsurfacefilterprop.h"
#include "packettypes/ntextui.h"
#include "packettypes/ntriangulationui.h"

#include <QPainter>
#include <QPixmap>
#include <QPlainTextEdit>

using namespace regina;

bool PacketManager::lockInitialised = false;
QImage PacketManager::lockSmall;
QImage PacketManager::lockBar;

QPixmap PacketManager::iconSmall(NPacket* packet, bool allowLock) {
    QString name = iconName(packet);
    if (name.isNull())
        return QPixmap();


    // TODO Does this work? emblem-locked and emblem aren't in the latest 
    // freedesktop spec.
    if (allowLock && ! packet->isPacketEditable()) {
        QPixmap overlay = ReginaSupport::themeIcon("emblem-locked").pixmap(8,8);
        QPixmap icon = ReginaSupport::regIcon(name).pixmap(16,16);
        QPixmap *result = new QPixmap(icon.width(),icon.height());
        result->fill(Qt::transparent);
        QPainter painter(result);
        painter.drawPixmap(0,0,icon);
        painter.drawPixmap(0,0,overlay);
        return *result; 
    }
    return ReginaSupport::regIcon(name).pixmap(16,16);
}

QPixmap PacketManager::iconBar(NPacket* packet, bool allowLock) {
    QString name = iconName(packet);
    if (name.isNull())
        return QPixmap();

    // TODO Sizes here. KDE says "use KIconLoader::Toolbar size" but I cannot
    // find a reference to what size that is, so using 16x16
    if (allowLock && ! packet->isPacketEditable()) {
        QPixmap overlay = ReginaSupport::themeIcon("emblem-locked").pixmap(8,8);
        QPixmap icon = ReginaSupport::regIcon(name).pixmap(16,16);
        QPixmap *result = new QPixmap(icon.width(),icon.height());
        result->fill(Qt::transparent);
        QPainter painter(result);
        painter.drawPixmap(0,0,icon);
        painter.drawPixmap(0,0,overlay);
        return *result; 
    }
    return ReginaSupport::regIcon(name).pixmap(16,16); 
}

PacketUI* PacketManager::createUI(regina::NPacket* packet,
        PacketPane* enclosingPane) {
    if (packet->getPacketType() == NAngleStructureList::packetType)
        return new NAngleStructureUI(dynamic_cast<NAngleStructureList*>(packet),
            enclosingPane);
    if (packet->getPacketType() == NContainer::packetType)
        return new NContainerUI(dynamic_cast<NContainer*>(packet),
            enclosingPane);
    if (packet->getPacketType() == NNormalSurfaceList::packetType)
        return new NNormalSurfaceUI(dynamic_cast<NNormalSurfaceList*>(packet),
            enclosingPane);
    if (packet->getPacketType() == NPDF::packetType)
        return new NPDFUI(dynamic_cast<NPDF*>(packet), enclosingPane);
    if (packet->getPacketType() == NScript::packetType) {
        return new NScriptUI(dynamic_cast<NScript*>(packet), enclosingPane);
    }
    if (packet->getPacketType() == NSurfaceFilter::packetType) {
        if (((NSurfaceFilter*)packet)->getFilterID() ==
                NSurfaceFilterCombination::filterID)
            return new NSurfaceFilterCombUI(
                dynamic_cast<NSurfaceFilterCombination*>(packet),
                enclosingPane);
        if (((NSurfaceFilter*)packet)->getFilterID() ==
                NSurfaceFilterProperties::filterID)
            return new NSurfaceFilterPropUI(
                dynamic_cast<NSurfaceFilterProperties*>(packet),
                enclosingPane);
        return new DefaultPacketUI(packet, enclosingPane);
    }
    if (packet->getPacketType() == NText::packetType) {
        return new NTextUI(dynamic_cast<NText*>(packet), enclosingPane);
    }
    if (packet->getPacketType() == NTriangulation::packetType)
        return new NTriangulationUI(dynamic_cast<NTriangulation*>(packet),
            enclosingPane);
    return new DefaultPacketUI(packet, enclosingPane);
}

QString PacketManager::iconName(NPacket* packet) {
    if (packet->getPacketType() == NAngleStructureList::packetType)
        return "packet_angles";
    if (packet->getPacketType() == NContainer::packetType)
        return "packet_container";
    if (packet->getPacketType() == NPDF::packetType)
        return "packet_pdf";
    if (packet->getPacketType() == NSurfaceFilter::packetType) {
        if (((NSurfaceFilter*)packet)->getFilterID() ==
                NSurfaceFilterCombination::filterID)
            return "filter_comb";
        else if (((NSurfaceFilter*)packet)->getFilterID() ==
                NSurfaceFilterProperties::filterID)
            return "filter_prop";
        else
            return "packet_filter";
    }
    if (packet->getPacketType() == NScript::packetType)
        return "packet_script";
    if (packet->getPacketType() == NNormalSurfaceList::packetType)
        return "packet_surfaces";
    if (packet->getPacketType() == NText::packetType)
        return "packet_text";
    if (packet->getPacketType() == NTriangulation::packetType)
        return "packet_triangulation";

    return QString();
}

File: swiftui/PacketManager.swift

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  Swift User Interface                                                  *
 *                                                                        *
 *  Copyright (c) 1999-2023, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  As an exception, when this program is distributed through (i) the     *
 *  App Store by Apple Inc.; (ii) the Mac App Store by Apple Inc.; or     *
 *  (iii) Google Play by Google Inc., then that store may impose any      *
 *  digital rights management, device limits and/or redistribution        *
 *  restrictions that are required by its terms of service.               *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public             *
 *  License along with this program; if not, write to the Free            *
 *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,       *
 *  MA 02110-1301, USA.                                                   *
 *                                                                        *
 **************************************************************************/

import SwiftUI
import ReginaEngine

struct PacketManager {
    /**
     * Returns a 32-point image representing the type of the given packet.
     */
    static func iconFor(_ packet: regina.SharedPacket) -> Image {
        switch packet.type() {
        case regina.PacketType.AngleStructures:
            return Image("Angles")
        case regina.PacketType.Attachment:
            return Image("Attachment")
        case regina.PacketType.Container:
            return Image("Container")
        case regina.PacketType.Link:
            return Image("Link")
        case regina.PacketType.NormalHypersurfaces:
            return Image("Hypersurfaces")
        case regina.PacketType.NormalSurfaces:
            return Image("Surfaces")
        case regina.PacketType.Script:
            return Image("Script")
        case regina.PacketType.SnapPea:
            return Image("SnapPea")
        case regina.PacketType.SurfaceFilter:
            return Image("Filter")
        case regina.PacketType.Text:
            return Image("Text")
        case regina.PacketType.Triangulation2:
            return Image("Triangulation2")
        case regina.PacketType.Triangulation3:
            return Image("Triangulation3")
        case regina.PacketType.Triangulation4:
            return Image("Triangulation4")
        case regina.PacketType.Triangulation5:
            return Image("Triangulation5")
        case regina.PacketType.Triangulation6:
            return Image("Triangulation6")
        case regina.PacketType.Triangulation7:
            return Image("Triangulation7")
        case regina.PacketType.Triangulation8:
            return Image("Triangulation8")
        // TODO: Cases 9..15, conditionally compiled
        default:
            // TODO: What do we return here?
            return Image("")
        }
    }
}
